### ISSUE TITLE
Can't make cli work 😭

### ISSUE THREAD

**Issue by JustACuteGirlLookingForAManToFixHerBugs** [beat-machine/beat-machine#53]
So, I get this error:

```
python -m beatmachine --input yourmom.mp3 --effects effects.json --output yourmom2.mp3
Traceback (most recent call last):
  File "C:\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Python39\lib\site-packages\beatmachine\__main__.py", line 20, in <module>
    main()
  File "C:\Python39\lib\site-packages\click\core.py", line 1134, in __call__
    return self.main(*args, **kwargs)
  File "C:\Python39\lib\site-packages\click\core.py", line 1059, in main
    rv = self.invoke(ctx)
  File "C:\Python39\lib\site-packages\click\core.py", line 1401, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Python39\lib\site-packages\click\core.py", line 767, in invoke
    return __callback(*args, **kwargs)
  File "C:\Python39\lib\site-packages\beatmachine\__main__.py", line 14, in main
    effects = [bm.effects.load_from_dict(e) for e in json.loads(effects)]
  File "C:\Python39\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Python39\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Python39\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

Here's my effects.json:
```
{
	"type": "swap",
	"x_period": 2, 
	"y_period": 4
	}
```
(your example)

Basically, I can't make it work no matter what.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
```
python -m beatmachine --input yourmom.mp3 --effects "{"type": "swap","x_period": 2, "y_period": 4}" --output yourmom2.mp3
Traceback (most recent call last):
  File "C:\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Python39\lib\site-packages\beatmachine\__main__.py", line 20, in <module>
    main()
  File "C:\Python39\lib\site-packages\click\core.py", line 1134, in __call__
    return self.main(*args, **kwargs)
  File "C:\Python39\lib\site-packages\click\core.py", line 1059, in main
    rv = self.invoke(ctx)
  File "C:\Python39\lib\site-packages\click\core.py", line 1401, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Python39\lib\site-packages\click\core.py", line 767, in invoke
    return __callback(*args, **kwargs)
  File "C:\Python39\lib\site-packages\beatmachine\__main__.py", line 14, in main
    effects = [bm.effects.load_from_dict(e) for e in json.loads(effects)]
  File "C:\Python39\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Python39\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Python39\lib\json\decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 2 (char 1)
```

This is what I get when not using effects.json

**Comment by branchpanic** [comment]
Try putting it in an array like in this example (note the extra `[` `]` which yours lacks):
```
$ python -m beatmachine -i in.mp3 -e '[{"type": "swap", "x_period": 2, "y_period": 4}]' -o out.mp3
```
I've recorded the need for a better error message in the overarching CLI redesign issue

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
I'm on Windows is that makes a difference.
```
D:\Cloud\Git\beat-machine>python -m beatmachine --input yourmom.mp3 --effects '[{"type": "swap","x_period": 2, "y_period": 4}]' --output yourmom2.mp3
usage: beatmachine [-h] [--version] [--input INPUT] [--effects EFFECTS] [--output OUTPUT] [--serialize] [--bpm BPM]
                   [--tolerance TOLERANCE]
beatmachine: error: unrecognized arguments: swap,x_period: 2, y_period: 4}]'

D:\Cloud\Git\beat-machine>python -m beatmachine --input yourmom.mp3 --effects '[{"type": "swap", "x_period": 2, "y_period": 4}]' --output yourmom2.mp3
usage: beatmachine [-h] [--version] [--input INPUT] [--effects EFFECTS] [--output OUTPUT] [--serialize] [--bpm BPM] [--tolerance TOLERANCE]
beatmachine: error: unrecognized arguments: swap, x_period: 2, y_period: 4}]'

D:\Cloud\Git\beat-machine>python -m beatmachine --input yourmom.mp3 --effects "[{"type": "swap", "x_period": 2, "y_period": 4}]" --output yourmom2.mp3
Traceback (most recent call last):
  File "C:\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\Cloud\Git\beat-machine\beatmachine\__main__.py", line 107, in <module>
    main()
  File "D:\Cloud\Git\beat-machine\beatmachine\__main__.py", line 43, in main
    effects_json = json.loads(args.effects)
  File "C:\Python39\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Python39\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Python39\lib\json\decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 3 (char 2)

```

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
```
D:\Cloud\Git\beat-machine>python -m beatmachine -i in.mp3 -e '[{"type": "swap", "x_period": 2, "y_period": 4}]' -o out.mp3
usage: beatmachine [-h] [--version] [--input INPUT] [--effects EFFECTS] [--output OUTPUT] [--serialize] [--bpm BPM] [--tolerance TOLERANCE]
beatmachine: error: unrecognized arguments: swap, x_period: 2, y_period: 4}]'
```

Here I just copied your command

**Comment by branchpanic** [comment]
I haven't used CMD in a while, but it looks like it escapes strings weirdly. Try putting the fixed effects string in a file like you were before:
```json
[
    {"type": "swap", "x_period": 2, "y_period": 4}
]
```
And using your first command:
```
python -m beatmachine --input yourmom.mp3 --effects effects.json --output yourmom2.mp3
```

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
What should I use instead of CMD?

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
I've made it work! OMG. That's why I don't like the CLIs - there are ALWAYS some errors flying around.

But the serialize thingie is just... not working. First, it won't run a command without --e effects.json. _Dura lex, sed lex_. And it doesn't like to eat a good old _non-existent file_, no, it prefers an empty file - with a slight touch of an empty array inside.

But then - and you gotta appreciate the irony - after eating a very real empty array, and then imitating hard work for a minute or two, it then proudly exclaims that it `Wrote output to in.beat` only to produce exactly nothing. 

That it, of course, refuses to eat then - because it's not even an empty file, there's just nothing.

Thank you, I hate it.

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
G in GUI stands for 'God's'

**Comment by JustACuteGirlLookingForAManToFixHerBugs** [comment]
[I Wish I Was Someone Better (Bitch I'm Better - I'm With Simon).mp3](https://bananas.s-ul.eu/p3x2JXKs)

**Comment by branchpanic** [comment]
Closing the issue now since it seems that it's "working" more or less, but I've noted your feedback about the buggy `--serialize` flag. This is already being addressed in #41 for what will hopefully be the last major release I have time to work on.

> What should I use instead of CMD?

I personally use PowerShell or zsh via WSL, but use whatever you're most comfortable with.
